motoko/ios-notifications/dapp-demo/src/www/src/auth.ts

```diff
@@ -0,0 +1,144 @@
+import { Identity, SignIdentity } from "@dfinity/agent";
+import { AuthClient, IdleOptions } from "@dfinity/auth-client";
+import {
+  AuthClientStorage,
+  IdbStorage,
+  KEY_STORAGE_DELEGATION,
+  KEY_STORAGE_KEY,
+} from "@dfinity/auth-client/lib/cjs/storage";
+
+export enum AuthLoginType {
+  Desktop,
+  Mobile = "callback_url",
+}
+
+export type IdentityParam = {
+    key?: string;
+    delegation?: string;
+}
+
+export class Auth {
+  private static identityQueryItem = "_identity";
+  private storage!: AuthClientStorage;
+  private authClient!: AuthClient;
+  private days = BigInt(1);
+  private hours = BigInt(24);
+  private nanoseconds = BigInt(3600000000000);
+
+  constructor(authClient: AuthClient, storage: AuthClientStorage) {
+    this.authClient = authClient;
+    this.storage = storage;
+  }
+
+  private static currentURL(): URL {
+    return new URL(window.location.href);
+  }
+
+  public login(onAuthenticated: (auth: Auth) => Promise<void>): void {
+    this.authClient.login({
+        onSuccess: async () => {
+            onAuthenticated(this);
+        },
+        identityProvider:
+            process.env.DFX_NETWORK === "ic"
+            ? "https://identity.ic0.app/#authorize"
+            : `http://localhost:${process.env.REPLICA_PORT}/?canisterId=${process.env.LOCAL_II_CANISTER}#authorize`,
+        // Maximum authorization expiration is 30 days
+        maxTimeToLive: this.days * this.hours * this.nanoseconds,
+    });
+  }
+
+  public client(): AuthClient {
+    return this.authClient;
+  }
+
+  public loginType(): AuthLoginType {
+    if (Auth.currentURL().searchParams.has(AuthLoginType.Mobile)) {
+      return AuthLoginType.Mobile;
+    }
+
+    return AuthLoginType.Desktop;
+  }
+
+  public static async create(options?: {
```

Comment:
This still creates the delegation keypair in the dapp by agent-js, right? I think it is fine this way for this example, but I would add an explanation somewhere that the private key could be kept in the secure enclave (generated by the native) app and the public key sent over to the web view (and the delegation back), which would be more secure an make the design resistant to XSS vulnerabilities in the dapp.

Reply:
Yes, it's still being created in the dapp. 

i've added the suggestion of generating the keypair from the native app in the security considerations section of the readme.